ArchiSteamFarm.Tests/Bot.cs

```diff
@@ -0,0 +1,197 @@
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 1 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 2), 1 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void MultipleSets() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1),
+				CreateCard(1),
+				CreateCard(2),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(items, cardsPerSet: 2, amountPerCard: 2);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 2 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 2), 2 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void MultipleSetsDifferentAmount() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, 2),
+				CreateCard(2),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(items, cardsPerSet: 2, amountPerCard: 2);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 2 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 2), 2 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void MoreCardsThanNeeded() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1),
+				CreateCard(1),
+				CreateCard(2),
+				CreateCard(3)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(items, cardsPerSet: 3, amountPerCard: 1);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 1 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 2), 1 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 3), 1 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void CardsWithOtherRarity() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, rarity: Steam.Asset.ERarity.Common),
+				CreateCard(1, rarity: Steam.Asset.ERarity.Rare)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(items, cardsPerSet: 1, amountPerCard: 1);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 1 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
```

Comment:
This test is actually wrong on two levels:
a) If what you've written was true, so you had two cards from two different sets for the same appID (one common and one rare), and the set has 1 card, then the result should be 2 sets with 1 card each.
b) As I said previously, you won't have a situation of the same class having different description (rarity in this case), because class identifies type of the item in unique fashion across combination of appID+contextID

Rewrite this test into three:

1. First verifying that the situation `class 1, amount 1, type common` + `class 2, amount 1, type common`, same other details, with set of 1 indeed generates **two** items to send, as those are two finished sets with varying rarity for the same appID.

2. Second verifying the situation with the same input as above but with 2 cards per set generates empty output.

3. Third being a mix of both, 2 items of one rarity, 3 items of other rarity and 4 items of yet another rarity, with 3 cards per set should generate 7 items total, 3 of other rarity and 4 of yet another rarity.

Reply:
And all of the above might require from you to fix the logic if it doesn't work that way (yet).

Reply:
about number 3: If we have 4 different class IDs but a set consists of 3 cards (according to badge page) we will not send items and I think the code segment we use in Bot.cs on line 3098 is correct in doing so, because if we sent 4 items, we would send more than is required for the badge.

Reply:
That's true, it doesn't make sense, so you can limit it to 2 3 2 or 2 3 3.

Reply:
I actually prefer to test  2,3,4 just to have all cases covered. But the expected result is 3 (As done in the tests from the last commit)